### PR TITLE
[Feature][Config] Support persistence of configuration items modified at runtime

### LINKED ISSUE #4703: [Proposal] Support persistence of configuration items modified at runtime
**Is your feature request related to a problem? Please describe.**

Doris now support modified FE and BE config at runtime. But the modification will be invalid after process restart.
This will make the cluster operation and maintenance more complicated, because if the process restarts unexpectedly,
the configuration changes made during runtime will become invalid and need to be modified again.

**Describe the solution you'd like**

I'd like to support persistence of configuration items modified at runtime. 
The modified configurations will be saved in `fe_custom.conf/be_custom.conf` under `conf/` dir.

And when process restarts, it will first load config from `fe.conf/be.conf`, then load `fe_custom.conf/be_custom.conf`.
The config in `xx_custom.conf` will overwrite the same config item in `fe.conf/be.conf`

### PR DESCRIPTION
**Pull request by morningman**
## Proposed changes

Support persistence of configuration items modified at runtime via HTTP API.

```
FE:
GET /api/_set_config?key=value&persist=true

BE
POST /api/update_config?key=value&persist=true
```

The modified config will be saved in `fe_custom.conf` or `be_custom.conf`.
And when process starts, it will load `fe.conf/be.conf` first, then `fe_custom.conf/be_custom.conf`.

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have create an issue on (Fix #4703), and have described the bug/feature there in detail
- [x] If this change need a document change, I have updated the document

### CODE REVIEW

File: be/src/common/configbase.cpp [hunk 1]
```diff
@@ -227,6 +231,34 @@ bool Properties::get(const char* key, const char* defstr, T& retval) const {
...
+    }
+    st = FileSystemUtil::create_file(conffile);
+    if (!st.ok()) {
+        return false;
+    }
+
+    std::ofstream out(conffile);
+    out << "# THIS IS AN AUTO GENERATED CONFIG FILE.\n";
+    out << "# You can modify this file manually, and the configurations in this file\n";
+    out << "# will overwrite the configurations in fe.conf\n";
```
Comment (acelyc111) [inline be/src/common/configbase.cpp:252]:
```suggestion
    out << "# will overwrite the configurations in be.conf\n";
```

File: be/src/http/action/update_config_action.cpp [hunk 2]
```diff
@@ -38,25 +38,53 @@ namespace doris {
...
     LOG(INFO) << req->debug_string();
 
     Status s;
     std::string msg;
-    if (req->params()->size() != 1) {
+    // We only support set one config at a time, and along with a optional param "persist".
+    // So the number of query params should at most be 2.
+    if (req->params()->size() > 2 || req->params()->size() < 1) {
         s = Status::InvalidArgument("");
         msg = "Now only support to set a single config once, via 'config_name=new_value'";
```
Comment (acelyc111) [inline be/src/http/action/update_config_action.cpp:52]: I think you should also describe the optional parameter `persist` in error message.

File: be/src/http/action/update_config_action.cpp [hunk 3]
```diff
@@ -38,25 +38,53 @@ namespace doris {
...
+                LOG(INFO) << "set_config " << config << "=" << new_value << " success";
+            } else {
+                LOG(WARNING) << "set_config " << config << "=" << new_value << " failed";
+                msg = strings::Substitute("set $0=$1 failed, reason: $2", config, new_value,
+                        s.to_string());
+            }
+        } else if (req->params()->size() == 2) {
+            if (req->params()->find(PERSIST_PARAM) == req->params()->end()) {
+                s = Status::InvalidArgument("");
+                msg = "Now only support to set a single config once, via 'config_name=new_value'";
```
Comment (acelyc111) [inline be/src/http/action/update_config_action.cpp:68]: Same

File: docs/en/administrator-guide/config/be_config.md [hunk 4]
```diff
@@ -30,51 +30,70 @@ under the License.
 
 This document mainly introduces the relevant configuration items of BE.
 
+The BE configuration file `be.conf` is usually stored in the `conf/` directory of the BE deployment path. In version 0.14, another configuration file `be_custom.conf` will be introduced. The configuration file is used to record the configuration items that are dynamically configured and persisted by the user during operation.
```
Comment (acelyc111) [inline docs/en/administrator-guide/config/be_config.md:33]: ... is used for recording ... ?

File: docs/en/administrator-guide/config/be_config.md [hunk 5]
```diff
@@ -30,51 +30,70 @@ under the License.
...
+    ```
+    curl -X POST http://{be_ip}:{be_http_port}/api/update_config?{key}={value}'
+    ```
 
-         ```curl -X POST http://{be_ip}:{be_http_port}/api/update_config?{key}={value}'```
+    In version 0.13 and before, the configuration items modified in this way will become invalid after the BE process restarts. In 0.14 and later versions, the modified configuration can be persisted through the following command. The modified configuration items are stored in the `be_custom.conf` file.
 
-         **Configuration items modified in this way will become invalid after the BE process restarts. **
+    ```
+    curl -X POST http://{be_ip}:{be_http_port}/api/update_config?{key}={value}&persis=true'
```
Comment (acelyc111) [inline docs/en/administrator-guide/config/be_config.md:62]:
```suggestion
    curl -X POST http://{be_ip}:{be_http_port}/api/update_config?{key}={value}&persist=true'
```

File: docs/zh-CN/administrator-guide/config/be_config.md [hunk 6]
```diff
@@ -45,9 +52,15 @@ BE 的配置项有两种方式进行配置：
...
-	```curl -X POST http://{be_ip}:{be_http_port}/api/update_config?{key}={value}'```
+	```
+	curl -X POST http://{be_ip}:{be_http_port}/api/update_config?{key}={value}'
+	```
 
-	**通过该方式修改的配置项将在 BE 进程重启后失效。**
+	在 0.13 版本及之前，通过该方式修改的配置项将在 BE 进程重启后失效。在 0.14 及之后版本中，可以通过以下命令持久化修改后的配置。修改后的配置项存储在 `be_custom.conf` 文件中。
+	
+	```
+	curl -X POST http://{be_ip}:{be_http_port}/api/update_config?{key}={value}&persis=true'
```
Comment (xy720) [inline docs/zh-CN/administrator-guide/config/be_config.md:62]:
```suggestion
	curl -X POST http://{be_ip}:{be_http_port}/api/update_config?{key}={value}&persist=true'
```

File: fe/fe-core/src/main/java/org/apache/doris/common/ConfigBase.java [hunk 7]
```diff
@@ -284,4 +308,31 @@ public synchronized static boolean checkIsMasterOnly(String key) {
...
+
+        Properties props = new Properties();
+        props.load(new FileReader(customConfFile));
+
+        for (Map.Entry<String, String> entry : customConf.entrySet()) {
+            props.setProperty(entry.getKey(), entry.getValue());
+        }
+
+        try (FileOutputStream fos = new FileOutputStream(file)) {
+            props.store(fos, "THIS IS AN AUTO GENERATED CONFIG FILE. DO NOT EDIT IT!\n\n" +
```
Comment (xy720) [inline fe/fe-core/src/main/java/org/apache/doris/common/ConfigBase.java:333]: I think users can edit this file if they need to.

File: be/src/common/configbase.cpp (hunk 1 above)
Comment (morningman) [reply be/src/common/configbase.cpp:252]: ok

File: be/src/http/action/update_config_action.cpp (hunk 2 above)
Comment (morningman) [reply be/src/http/action/update_config_action.cpp:52]: ok

File: be/src/http/action/update_config_action.cpp (hunk 3 above)
Comment (morningman) [reply be/src/http/action/update_config_action.cpp:68]: ok

File: docs/en/administrator-guide/config/be_config.md (hunk 4 above)
Comment (morningman) [reply docs/en/administrator-guide/config/be_config.md:33]: This is translated by Google Translator... So, I think it is correct ^^

File: docs/zh-CN/administrator-guide/config/be_config.md (hunk 6 above)
Comment (morningman) [reply docs/zh-CN/administrator-guide/config/be_config.md:62]: OK

File: fe/fe-core/src/main/java/org/apache/doris/common/ConfigBase.java (hunk 7 above)
Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/common/ConfigBase.java:333]: OK

**Review by yangzhg — DISMISSED** [review]
+1

**Review by yangzhg — APPROVED** [review]
+1

**Pull request merged**